library/alloc/src/boxed/thin.rs

```diff
@@ -91,6 +95,74 @@ impl<T> ThinBox<T> {
 
 #[unstable(feature = "thin_box", issue = "92791")]
 impl<Dyn: ?Sized> ThinBox<Dyn> {
+    #[cfg(not(no_global_oom_handling))]
+    fn new_unsize_zst<T>(value: T) -> Self
+    where
+        T: Unsize<Dyn>,
+    {
+        assert!(mem::size_of::<T>() == 0);
+
+        // Utility to statically allocate metadata followed by ZST value.
+        struct DynZstAlloc<T, Dyn: ?Sized>(PhantomData<(T, Dyn)>);
+
+        impl<'a, T: 'a, Dyn: ?Sized + 'a> DynZstAlloc<T, Dyn>
+        where
+            T: Unsize<Dyn>,
+        {
+            // We are returning `[T; 0]` here instead of `T` because
+            // compiler instantiates this code even for non-ZST,
+            // and reports a pointer going beyond the allocation.
+            const ALLOC: &'a [T; 0] = {
```

Comment:
Looks like that type only exists only to have an associated const? In that case you can probably use an inline const instead.

Reply:
Much simpler. Thanks!